added Polimorphic Projections

JanEggers opened the pull request:
I consider this a draft that does what I need.

If the design is approved I will also create some tests.

![image](https://user-images.githubusercontent.com/6747376/229468098-15fbe46f-8743-4cc9-ac18-36d2c523ef20.png)

![image](https://user-images.githubusercontent.com/6747376/229468241-ee017733-2c5b-47df-8506-7d92cbc4b682.png)

new queries contain the expected subtypes.

fixes #4271

JanEggers commented:
![image](https://user-images.githubusercontent.com/6747376/229483003-08b77c2f-f52b-47a7-bd8d-7317912352d5.png)

here is also the generated expression tree typically the lambda would only contain the last else path that creates the base class

lbargaoanu commented:
We have [this](https://github.com/AutoMapper/AutoMapper/blob/73ba4003e49e37e5fffec49b67e7d9ea1cbb8a8a/src/UnitTests/Projection/ProjectionMappers.cs) extension point. I suggest that you try to implement it outside AM while we discuss things.

lbargaoanu commented:
Moved from the discussion
@lbargaoanu tried your suggestion in the pr. it may be possible but not viable. I would have to replace calling into ProjectionBuilder.CreateProjection to replace CreateDestination to be able to change the new expression. that way i have to effectivly copy all the code in Projectionbuilder. 

maybe you could improve the design so GetProjectionMapper is also called by EnumerableProjectionMapper. right now EnumerableProjectionMapper does not call GetProjectionMapper making it hard to create a custom mapper for elements inside a list.

lbargaoanu commented:
Did you try calling `CreateProjection` for each of the included types?

JanEggers commented:
yes. but i only get 3 calls for IsMatch Parent.Id, Parent.Children, Child.Id.

but i dont get a call for Parent.Children ElementType aka ChildModel  => ChildDto

lbargaoanu commented:
Whoever handles that, if you're before them in the list, you can take over.

JanEggers commented:
tried again. its 60 lines of code and im still missing the equivalent of ProjectProperties.

I think this should be in the box

https://github.com/JanEggers/GraphqlRepro/blob/2e32c3a29003c8eada2124a22d1207c485d31e54/GraphQlRepro/Program.cs#L69-L124

lbargaoanu commented:
I've looked a little deeper and is [used](https://github.com/AutoMapper/AutoMapper/commit/5b0cf17a3b19dcbd7ec214efb96455309c330643) to be the case that the list of projection mappers was checked on every projection request. If that helps, we'll consider bringing it back.

JanEggers commented:
for me it is not viable implementing a custom projection mapper. it is less code to just manually write the expression.

lbargaoanu commented:
Then that's the way to go.

lbargaoanu commented:
How would this work with our let clause? Examples: https://github.com/AutoMapper/AutoMapper/blob/master/src/IntegrationTests/CustomMapFrom/MapObjectPropertyFromSubQuery.cs and some and the ones using methods from https://github.com/AutoMapper/AutoMapper/blob/master/src/IntegrationTests/IncludeMembers.cs.

JanEggers commented:
I dont understand the question. what is a let clause? 

the tests are all green on my branch so I asume everything is working fine.

can you maybe create a test of what you dont expect to work?

JanEggers commented:
I understand the problem now. the let will introduce an intermediate projection that looses the type information. I will try to rewrite the projection in a way that preserves the type information.

JanEggers commented:
I updated all the tests in IncludeMembers to also use inhertance.

I also fixed the let clause be including the whole source in the intermediate projection so the type information is not lost.

lbargaoanu commented:
That defeats the very purpose of `ProjectTo`, fetching only the required data to fill the dto.

JanEggers commented:
> That defeats the very purpose of `ProjectTo`, fetching only the required data to fill the dto.

nope its just an intermidiate projection that is discarded. only the columns present in the final select are fetched.

lbargaoanu commented:
As far as I can tell, that's correct. Revert all changes to files other than ProjectionBuilder. Don't use `IncludeAllDerived`, explicitly include the required maps.
Add new tests (with `CreateMap`) for both files I mentioned that use inheritance. Add at least one an extra column to the derived classes and assert they have the expected data.
When that is done, I'll take a look at the tests to try and see if they work as expected.
After that we can discuss the implementation.

JanEggers commented:
> Revert all changes to files other than ProjectionBuilder

can you elaborate why? I was under the impression that if inheritance is supported by projection it should be available by AddProjection?

lbargaoanu commented:
I feel it's a niche use case and I believe the majority of people can safely ignore it exists. The rest can just use `CreateMap` or an `Advanced` extension method that casts `CreateProjection` to `IMappingExpression`.

lbargaoanu commented:
A thing I forgot to mention is that I have a branch ahead of master that surely conflicts with this one. I think we'll merge that one soon enough and you'll have to rebase. Not exactly major changes, but conflicts nevertheless.

lbargaoanu commented:
Constructor mapping has to work too. Some tests are [here](https://github.com/AutoMapper/AutoMapper/blob/master/src/UnitTests/Projection/ConstructorTests.cs).

JanEggers commented:
I updated the tests in [MapObjectPropertyFromSubQuery](https://github.com/AutoMapper/AutoMapper/blob/master/src/IntegrationTests/CustomMapFrom/MapObjectPropertyFromSubQuery.cs) and did the rebase

two of the tests are failing because there are now multiple FirstOrDefault calls. I updated the code to reuse the baseclass expressions for the derived classes but the issue persists. any advice whats wrong or how to fix it is very welcome.

also i noticed when mixing addprojection and addmapping I get an exception while projecting. I dont know if this is by design.

I will check / extend the constructor tests

lbargaoanu commented on src/IntegrationTests/IncludeMembers.cs:
```diff
@@ -920,22 +1536,45 @@ public class Destination
+            .Include<SourceA, DestinationA>()
+            .Include<SourceB, DestinationB>();
+        cfg.CreateMap<SourceA, DestinationA>();
+        cfg.CreateMap<SourceB, DestinationB>();
```
None of these derived maps actually have included members. I want to see at least two such cases, one for a simple included member, chained properties, and one that uses the let clause, such as this one, with a method. 
Check [the execution plan](https://docs.automapper.org/en/latest/Understanding-your-mapping.html). That extension is very useful :)

lbargaoanu replied:
Unless I've missed smth, you didn't address this and a similar comment for the subquery test.

JanEggers replied:
yes will do today. was busy with the family over the weekend :)

JanEggers replied:
I extended the tests. basicaly doing the same thing as for the base class in the derived class. that uncovered some problems with the current approach. I was able to fix some of the issues but there is at least one more issues remaining.  im not quite certain whats the issue as the expression it self looks ok but the entity framework transformation of the query cannot be translated to sql because it contains a a lamda that should probably be a property expression.

JanEggers replied:
my current thinking is that it would be better to have multiple intermediate let statements that would also improve expression caching.

e.g:

instead of:
```c#
.Select(s => new {
    Source = s,
    Inner = s.InnerSources.FirstOrDefault(),
    Inner#Details = s.InnerSources.FirstOrDefault().InnerSourceDetails.FirstOrDefault().
})
```
generate:
```c#
.Select(s => new {
    Source = s,
    Inner = s.InnerSources.FirstOrDefault()
}).Select(s => new {
    s.Source,
    s.Inner,
    Details = s.Inner.InnerSourceDetails.FirstOrDefault()
})
```

lbargaoanu replied:
One let clause should be enough for everyone :) It's not clear to me why EF would not handle the first expression. Except that it's too complicated. And if that's the case, just simplify your test so things don't compose. But of course I still want to see a subquery with a method in a derived class.

JanEggers replied:
im investigating

lbargaoanu commented:
Yes, don't mix `CreateMap` with `CreateProjection`. About the failing test, no hurry, it can wait. We'll figure it out eventually.

lbargaoanu commented on src/IntegrationTests/CustomMapFrom/MapObjectPropertyFromSubQuery.cs:
```diff
@@ -689,27 +1007,43 @@ public class DatabaseInitializer : DropCreateDatabaseAlways<ClientContext>
+            .Include<ProductArticleA, ProductArticleAModel>()
+            .Include<ProductArticleB, ProductArticleBModel>();
+        cfg.CreateMap<ProductArticleA, ProductArticleAModel>();
+        cfg.CreateMap<ProductArticleB, ProductArticleBModel>();
```
Similar to `IncludeMembers`, at least a few of these tests should have some configuration in the derived map. Certainly a subquery using the let clause and perhaps some random `MapFrom` to see that it's used.

lbargaoanu commented on src/IntegrationTests/IncludeMembers.cs:
```diff
@@ -1630,4 +1630,105 @@ public void Should_flatten()
         }
     }
+}
+public class IncludeOnlySelectedMembers : IntegrationTest<IncludeOnlySelectedMembers.DatabaseInitializer>
```
This needs to use an expression with `FirstOrDefault` to use the let clause.

JanEggers replied:
I changed the test like you requested locally but it is broken, although not a regression. the FirstOrDefault let expression brings in the whole object instead of the requested properties. 

I am investigating if this can be fixed by adding a projection that only includes the required properties

JanEggers replied:
found a solution

lbargaoanu replied:
But that just brings back the one piece of code you removed.

JanEggers replied:
the original code was not preserving type information. 
if you have suggestions on what code should be restored please let me know. then i will try to use that code instead.

JanEggers replied:
also i tried to repro on master but strangely there the query is correct even if the let clause contains the whole object included by firstOrDefault. im still investigating whats the difference there...

JanEggers replied:
I was able to repro the issue.
as soon as the let statement contains a conditional statement entity framework selects all the properties instead of the required properties.

https://github.com/JanEggers/AutoMapper/tree/queryContainsMoreThanRequiredProperties

JanEggers commented on src/UnitTests/Projection/ConstructorTests.cs:
```diff
@@ -65,51 +124,175 @@ class Source
+        cfg.CreateMap<Source, DestinationA>();
+        cfg.CreateMap<Source, DestinationB>();
+        cfg.CreateMap<SourceA, DestinationA>();
+        cfg.CreateMap<SourceB, DestinationB>();
```
does nor work either. for some reason it converts SourceA and SourceB toString instead of using the Properties on the objects to call the constructor

JanEggers replied:
changed member names now this test is green

JanEggers commented:
finally found a solution for the tests failing because ef was not not able to translate expressions.
the issue was caused by incorrect chaining of conditional source expressions. I am not 100% certain its the correct way to handle it as it is quite complicated to wrap your head around the expression trees....

still two failing tests remaining

lbargaoanu commented on src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs:
```diff
@@ -243,16 +297,43 @@ public override QueryExpressions GetSubQueryExpression(ProjectionBuilder builder
+                        {
+                            if (letMapInfos[j].MapFromSource.Type == letMapInfo.Marker.Type)
+                            {
+                                letMapInfos[j] = letMapInfos[j] with { MapFromSource = source };
```
The chaining should happen in `GetSourceExpression`, not here.

JanEggers replied:
I changed it on purpose as a resultig expression of one let property can be a source of another property. im not sure how to  code this recursion when the code is in  GetSourceExpression

JanEggers replied:
understood

JanEggers replied:
I was able to simplify code and reuse it some more let me know what you think...

JanEggers commented on src/UnitTests/Projection/ConstructorTests.cs:
```diff
@@ -43,20 +63,59 @@ public DestinationItem(DestinationValue destinationValue)
+            .Include<SourceA, DestinationA>()
+            .Include<SourceB, DestinationB>();
+        cfg.CreateMap<SourceA, DestinationA>()
+            .ForCtorParam("item", o => o.MapFrom(s => s.Items.FirstOrDefault())) //TODO: required for CTOR Validation but should not be required and messes up mapping
```
I changed the let property names to include the source types, this way there is no issue in having the same properties for different source types. the issue was that there have been 3 item properties due to the redundand constructor mappings.

lbargaoanu replied:
That's arbitrary. It should work in the same way as it does for properties. By building a path within the query that says how we got to that point.

JanEggers replied:
the problem is that it is an expression with a parameter. the new code checks the type of the parameter and creates a convert expression to the type of the parameter. and that results in 3 different source expressions that access the same property.

if the 3 properties should be treated as the same property I need to change the code to not look at the type of the parameter but look at the declaring type of the first memberexpression. 

is this a better solution?

lbargaoanu replied:
That cast is not needed in this case. Maybe try to have it only when it's actually needed, perhaps when accessing a member of a derived type.

JanEggers replied:
done

JanEggers commented:
tests are green now on my system let me know if I should add more tests change something in the implementation

lbargaoanu commented:
I see

  Message: 
System.NullReferenceException : Object reference not set to an instance of an object.

  Stack Trace: 
ConstructorLetClause.Should_construct_correctly() line 111

lbargaoanu commented on src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs:
```diff
@@ -165,11 +165,10 @@ Expression MapFromExpression(LambdaExpression mapFrom)
-                            parmaterExtractor.Visit(mapFrom.Body);
-                            if (parmaterExtractor.Parameter != null)
+                            var member = MemberAccessExpressionExtractorVisitor.Extract(mapFrom.Body, mapFrom.Parameters.First());
+                            if (member != null)
```
What does this actually mean? Why would accessing smth on the parameter matter?

JanEggers replied:
the code needs to decide weather a type conversion is required.

before I used the parameter type to detect this.

but that resulted in unnessesay conversions in case the parameter is of the derived type but the used property is defined in the base class.  

the new code only adds a conversion if the member of the parameter is actually declared on the derived type

JanEggers replied:
Ahh I was not thinking about using multiple members in one expression tree I will update the code

JanEggers replied:
done. i now extract all member expressions on the given parameter and use the most derived type to determine weather a conversion is required.

lbargaoanu commented:
I think it's fair to say we're not trending in the right direction :)
I think we should simply not mix inheritance with the let clause. So have everything inline in a single `Select`.
Just leave this branch for future reference and start over. Hopefully the new approach can work with less complexity.

JanEggers commented:
> So have everything inline in a single Select

I tried that suggestion but removing the let clause creates much bigger queries as the subqueries are now longer reused.

to clarify did you mean i should remove the let clause completely or just avoid using it when dealing with inheritance?

lbargaoanu commented:
Just avoid using it with inherintance, as is the case now for chained properties expressions.
And yes, the resulting queries won't look good, but at least the code generating them would be manageable.
